transports/websocket-websys/src/web_context.rs

```diff
@@ -1,57 +1,185 @@
-use wasm_bindgen::{prelude::*, JsCast};
+use std::fmt::Debug;
+use std::sync::atomic::{AtomicU8, Ordering};
+
+use wasm_bindgen::prelude::*;
+use wasm_bindgen::JsCast;
 use web_sys::window;
 
+#[wasm_bindgen]
+extern "C" {
+    #[derive(Debug, Clone)]
+    pub(super) type Global;
+
+    #[wasm_bindgen(method, getter, js_name = WorkerGlobalScope)]
+    fn worker(this: &Global) -> JsValue;
+
+    #[wasm_bindgen(method, catch, variadic, js_name = setInterval)]
+    fn set_interval(
+        this: &Global,
+        handler: &js_sys::Function,
+        timeout: i32,
+        arguments: &js_sys::Array,
+    ) -> Result<JsValue, JsValue>;
+
+    #[wasm_bindgen(method, catch, js_name = clearInterval)]
+    fn clear_interval(this: &Global, handle: JsValue) -> Result<(), JsValue>;
+}
+
+#[derive(Debug)]
+enum HandleValue {
+    Numeric(i32),
+    Opaque(JsValue),
+}
+
+// we're using additional indirection here, to prevent the caller from knowing
+// what type of handle is being returned.
+// we also don't implement `From<i32>` or `From<JsValue>` for this type to
+// circumvent the creation of invalid handles outside of a context.
+#[derive(Debug)]
+pub(crate) struct IntervalHandle(HandleValue);
+
+impl IntervalHandle {
+    fn from_i32(value: i32) -> Self {
+        Self(HandleValue::Numeric(value))
+    }
+
+    fn from_value(value: JsValue) -> Self {
+        Self(HandleValue::Opaque(value))
+    }
+
+    fn into_i32(self) -> Option<i32> {
+        match self.0 {
+            HandleValue::Numeric(value) => Some(value),
+            HandleValue::Opaque(value) => value.as_f64().map(|value| value as i32),
+        }
+    }
+
+    fn into_value(self) -> JsValue {
+        match self.0 {
+            HandleValue::Numeric(value) => JsValue::from_f64(value as f64),
+            HandleValue::Opaque(value) => value,
+        }
+    }
+}
+
+fn has(value: &JsValue, key: &str) -> bool {
+    js_sys::Reflect::has(value, &JsValue::from_str(key)).unwrap_or(false)
+}
+
+/// A wrapper around a `u8` that is used to indicate whether the `WebSocket`
+/// API is supported.
+///
+/// * 0x00: Not yet checked
+/// * 0x01: Supported
+/// * 0x02: Not supported
+/// * other: Unknown -> fallback to `0x00`
+struct WebSocketSupport(AtomicU8);
```

Comment:
Why isn't this just an enum?

Reply:
Using an `AtomicU8` instead allows us to load/store on &self, making it suitable for static variable use.

Reply:
I see, lets document this please.